File: benchmarks/livermore/k02_iccg.c
/* K2 — ICCG excerpt (1-based indexing, matches Fortran exactly)
   Fortran: DIMENSION X(1001), V(1001), N=101
   Halving do-while with stride-2 inner loop */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     101
#define NREPS 10000

int main(void) {
    double x[1002], v[1002];

    signel(x+1, 1001);
    signel(v+1, 1001);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    for (long rep = 0; rep < NREPS; rep++) {
        signel(x+1, 1001);
        long ii = N;
        long ipntp = 0;
        do {
            long ipnt = ipntp;
            ipntp = ipntp + ii;
            ii = ii / 2;
            long i = ipntp;
            for (long k = ipnt + 2; k <= ipntp; k += 2) {
                i = i + 1;
                x[i] = x[k] - v[k] * x[k-1] - v[k+1] * x[k+1];
            }
        } while (ii > 0);
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("x[1] = %f\n", x[1]);
    return 0;
}

File: benchmarks/livermore/k02_iccg.f
C     Kernel 2 -- ICCG excerpt (Incomplete Cholesky Conjugate Gradient)
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C
      PROGRAM K02
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP
      DIMENSION X(1001), V(1001)
      INTEGER IPNT, IPNTP, II
C
C     Initialize arrays using SIGNEL pattern
C
      SCALED = 0.1D0
C
C     Initialize V
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 10 K = 1, 1001
          BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
          FUZZ = -FUZZ
          V(K) = (BUZZ - FIZZ) * SCALED
   10 CONTINUE
C
      N = 101
      DO 200 REP = 1, 11495000
C         Reset X each rep
          FUZZ  = 1.2345D-3
          BUZZ  = 1.0D0 + FUZZ
          FIZZ  = 1.1D0 * FUZZ
          DO 20 K = 1, 1001
              BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
              FUZZ = -FUZZ
              X(K) = (BUZZ - FIZZ) * SCALED
   20     CONTINUE
C
          II = N
          IPNTP = 0
  100     CONTINUE
              IPNT = IPNTP
              IPNTP = IPNTP + II
              II = II / 2
              I = IPNTP
              DO 90 K = IPNT + 2, IPNTP, 2
                  I = I + 1
                  X(I) = X(K) - V(K) * X(K-1) - V(K+1) * X(K+1)
   90         CONTINUE
          IF (II .GT. 0) GO TO 100
  200 CONTINUE
C
      WRITE(*,900) X(1)
  900 FORMAT('K2 ICCG: x(1) = ', E25.15)
      END
